lib/CMakeLists.txt

```diff
@@ -6,8 +6,8 @@ add_compile_options($<TARGET_PROPERTY:compiler,warning_shadow_variables>)
 add_subdirectory(crc)
 if(NOT CONFIG_EXTERNAL_LIBC)
 add_subdirectory(libc)
-add_subdirectory(posix)
 endif()
+add_subdirectory(posix)
```

Comment:
There is a fundamental incompatibility between POSIX_API and using the host libC (EXTERNAL_LIBC).
Moving this out of the if make it easier for users/developers to enable that combination in the build which will lead to more confusion.
I would have this include remain conditional CONFIG_EXTERNAL_LIBC or {to COFIG_POSIX_API and make the POSIX_API kconfig depend also on !EXTERNAL_LIBC}.

Reply:
> There is a fundamental incompatibility between POSIX_API and using the host libC (EXTERNAL_LIBC).

I would disagree here.

But please, explain how you feel there is a fundamental incompatibility.

> Moving this out of the if make it easier for users/developers to enable that combination in the build which will lead to more confusion.

I also would disagree here.

But please, explain how you feel this would be confusing.

> I would have this include remain conditional CONFIG_EXTERNAL_LIBC or {to COFIG_POSIX_API and make the POSIX_API kconfig depend also on !EXTERNAL_LIBC}.

That's both unnecessary and contrary to the direction approved by the TSC / Architecture WG.

It also sounds vaguely like how CONFIG_POSIX_API used to be incompatible  with CONFIG_ARCH_POSIX, but just wrapped in different packaging.

That's fixed now though, isn't it?

At least that's what I and many others were led to believe  [here](https://github.com/zephyrproject-rtos/zephyr/issues/45100#issuecomment-1628397622).

Reply:
EXTERNAL_LIBC means the host C library (normally glibc). The host C library implements the POSIX API in itself, which for the OS related logic calls into the Linux kernel.
If you build Zephyr's POSIX compatibility layer (CONFIG_POSIX_API) with it you are going to find that many of its headers conflict with the host C library, and that you are building functions and declaring variables which are already provided by the host C library. If things are built without a lot of care, the result is that most symbols (functions and global variables) come from the host C library.
For a short period of time (2017-18) the combination was allowed in Zephyr, but that resulted in too much confusion and frustration, so we stopped supporting it.

The thing with native_sim is that it allows to build the embedded code with an embedded C library. That is why you can select CONFIG_POSIX_API with it. When you select CONFIG_POSIX_API for native_sim PICOLIBC is selected by default, and things just work.

If a developer forces a build in native_sim with CONFIG_EXTERNAL_LIBC and CONFIG_POSIX_API (which by default is prevented, but can be enabled by doing things like moving this line in question out of the `if`, between other changes), that developer would be forcefully replicating the native_posix incompatibility in native_sim.

Reply:
> EXTERNAL_LIBC means the host C library (normally glibc).

It is defined as "build with an external/user-provided libc", so in the context of arch posix, that would mean the host libc (typically), but really it is any user-provided libc, which *does* include those which do not have posix support built-in.

So yes, POSIX_API is not mutually exclusive with `EXTERNAL_LIBC`.

> The host C library implements the POSIX API in itself, which for the OS related logic calls into the Linux kernel.

> If you build Zephyr's POSIX compatibility layer (CONFIG_POSIX_API) with it you are going to find that many of its headers conflict with the host C library, and that you are building functions and declaring variables which are already provided by the host C library. If things are built without a lot of care, the result is that most symbols (functions and global variables) come from the host C library.

Mostly agree.

> For a short period of time (2017-18) the combination was allowed in Zephyr, but that resulted in too much confusion and frustration, so we stopped supporting it.

Yes, that was bad, and it was good that we stopped it.

> The thing with native_sim is that it allows to build the embedded code with an embedded C library. That is why you can select CONFIG_POSIX_API with it. When you select CONFIG_POSIX_API for native_sim PICOLIBC is selected by default, and things just work.

Right - no issues there.

> If a developer forces a build in native_sim with CONFIG_EXTERNAL_LIBC and CONFIG_POSIX_API (which by default is prevented)

The above is probably not even an issue in this PR anymore, if we have concesus that _application_ conformance symbols are only ever locally defined for the library, which you suggested.

So I'm hoping you are still agreeable to that, and this issue can be resolved.

But, on the topic of external libc vs posix api..

Technically, anything that happens "below the line" (i.e. as part of the implementation of native_sim) can (should?) be mostly independent of what happens "above the line" (in the Zephyr application).

The underlying (and still very much present) problem is due to C's global namespace. The names of the calls that are made by native_sim conflict with names of the calls that the zephyr application can make, and we do not want the two to mix.

If that weren't an issue, then I don't believe there would be a reason to make posix api and external libc mutually exclusive.

Would you agree?

Reply:
> It is defined as "build with an external/user-provided libc" ... but really it is any user-provided libc

That option help text was overly ambitious compared with what was implemented, like it happens some times.
Keith was kind enough to add CONFIG_NATIVE_LIBC recently, so the filtering can be done by that instead of EXTERNAL_LIBC if you prefer.

> The names of the calls that are made by native_sim conflict with names of the calls that the zephyr application can make, and we do not want the two to mix.

The C library native_sim uses is *not* exposed to the embedded code. That is the whole point. But when you select CONFIG_EXTERNAL_LIBC in Zephyr you are actively choosing to build the embedded code, include the POSIX API compatibility shim code with the host libC. So you cannot expect that same code to not see the host libC, because you just chose it.

Reply:
Using `if(NOT CONFIG_NATIVE_LIBC)`